### ISSUE TITLE
Support for GraphQL multipart request 

### ISSUE THREAD

**Issue by srekapalli** [micronaut-projects/micronaut-graphql#12]
Trying to make GraphQL multipart request work as per this specification [https://github.com/jaydenseric/graphql-multipart-request-spec](https://github.com/jaydenseric/graphql-multipart-request-spec) , but receiving error when I send a multipart request. Is Multipart request supported now in Micronaut/GraphQL extension ?

Error:
**/string - Failed to convert argument [string] for value [null] due to: Cannot deserialize instance of `java.lang.String` out of START_OBJECT token
 at [Source: UNKNOWN; line: -1, column: -1]] using codec: io.micronaut.jackson.codec.JsonMediaTypeCodec@439d545c**

---- Logs ----
```
21:52:10.675 [nioEventLoopGroup-1-8] DEBUG i.m.h.s.netty.RoutingInBoundHandler - Matching route POST - /graphql
21:52:10.675 [nioEventLoopGroup-1-8] DEBUG i.m.h.s.netty.RoutingInBoundHandler - Matched route POST - /graphql to controller class io.micronaut.configuration.graphql.GraphQLController
21:52:10.676 [nioEventLoopGroup-1-8] DEBUG i.m.context.DefaultBeanContext - Resolved existing bean [io.micronaut.http.server.netty.FormDataHttpContentSubscriberFactory@5eeaaf88] for type [interface io.micronaut.http.server.netty.HttpContentSubscriberFactory] and qualifier [Content-Type: multipart/form-data;boundary=------------------------e67f840163e3fadc]
21:52:10.678 [nioEventLoopGroup-1-8] DEBUG i.m.c.beans.DefaultBeanIntrospector - No BeanIntrospection found for bean type: class java.lang.String
21:52:10.678 [nioEventLoopGroup-1-8] DEBUG io.micronaut.core.reflect.ClassUtils - Cannot instantiate type [class java.lang.String] without reflection. Attempting reflective instantiation
21:52:10.678 [nioEventLoopGroup-1-8] DEBUG i.m.context.DefaultBeanContext - Resolved existing bean [io.micronaut.jackson.bind.JacksonBeanPropertyBinder@69626e28] for type [interface io.micronaut.core.bind.BeanPropertyBinder] and qualifier [null]
21:52:10.678 [nioEventLoopGroup-1-8] DEBUG i.m.context.DefaultBeanContext - Registering singleton bean io.micronaut.http.server.exceptions.ConversionErrorHandler@246a6496 for type [<ConversionErrorException,Object> io.micronaut.http.server.exceptions.ExceptionHandler] using bean key <ConversionErrorException,Object> io.micronaut.http.server.exceptions.ConversionErrorHandler
21:52:10.678 [nioEventLoopGroup-1-8] DEBUG i.m.context.DefaultBeanContext - Registering singleton bean io.micronaut.http.server.exceptions.ConversionErrorHandler@246a6496 for type [<ConversionErrorException,Object> io.micronaut.http.server.exceptions.ExceptionHandler] using bean key <ConversionErrorException,Object> io.micronaut.http.server.exceptions.ConversionErrorHandler
21:52:10.679 [nioEventLoopGroup-1-8] DEBUG i.m.context.DefaultBeanContext - Registering singleton bean io.micronaut.http.server.exceptions.ConversionErrorHandler@246a6496 for type [<ConversionErrorException,Object> io.micronaut.http.server.exceptions.ExceptionHandler] using bean key <ConversionErrorException,Object> io.micronaut.http.server.exceptions.ConversionErrorHandler
21:52:10.679 [nioEventLoopGroup-1-8] DEBUG i.m.context.DefaultBeanContext - Resolved existing bean [io.micronaut.http.server.exceptions.ConversionErrorHandler@246a6496] for type [interface io.micronaut.http.server.exceptions.ExceptionHandler] and qualifier [<ConversionErrorException,Object>]
21:52:10.679 [nioEventLoopGroup-1-8] DEBUG i.m.h.s.netty.RoutingInBoundHandler - Encoding emitted response object [ /string - Failed to convert argument [string] for value [null] due to: Cannot deserialize instance of `java.lang.String` out of START_OBJECT token
 at [Source: UNKNOWN; line: -1, column: -1]] using codec: io.micronaut.jackson.codec.JsonMediaTypeCodec@439d545c
```

**Comment by marceloverdijk** [comment]
Hi @srekapalli, multipart requests are indeed not (yet) supported. 
Currently only the [Serving over HTTP](https://graphql.org/learn/serving-over-http/) specs are implemented.

I will look into this and consider supporting the multipart spec. Do you know if other graphql implementations implement this https://github.com/jaydenseric/graphql-multipart-request-spec spec?

**Comment by sflahave** [comment]
@marceloverdijk - It looks like GraphQL-Java-Kickstart added support for that spec: 
https://github.com/graphql-java-kickstart/graphql-java-servlet/pull/102

**Comment by marceloverdijk** [comment]
Correct [graphql-java-kickstart/graphql-java-servlet](https://github.com/graphql-java-kickstart/graphql-java-servlet) does indeed support it. I was wondering if other implementations like graphqljs or the express module also support it or not, to see if that spec is widely adopted.

**Comment by marceloverdijk** [comment]
After initial investigation Micronaut currently does not support retrieving the parts of multipart request. At least not in the sense needed to implement the https://github.com/jaydenseric/graphql-multipart-request-spec. 

Micronaut does support retrieving parts, but only if the part names are known upfront, which we don't know for https://github.com/jaydenseric/graphql-multipart-request-spec.

https://github.com/micronaut-projects/micronaut-core/issues/1957 is created to provide a way to retrieve the parts. Until then https://github.com/jaydenseric/graphql-multipart-request-spec cannot be implemented unfortunately.

**Comment by marceloverdijk** [comment]
Support for retrieving multiparts is scheduled for Micronaut Core `1.3.0` (see https://github.com/micronaut-projects/micronaut-core/issues/1957). We have to wait for that before we can implement https://github.com/jaydenseric/graphql-multipart-request-spec in `micronaut-graphql`.

**Comment by marceloverdijk** [comment]
Support for [retrieving multiparts](https://github.com/micronaut-projects/micronaut-core/issues/1957) is now added to Micronaut Core and will be part of Micronaut `1.3.0`.

However, there is another Micronaut Core issue pending to implement this feature in Micronaut GraphqL: https://github.com/micronaut-projects/micronaut-core/issues/1971 as we currently cannot distinguish the routes.

**Comment by marceloverdijk** [comment]
Note https://github.com/micronaut-projects/micronaut-core/issues/1971 has been released as part of Micronaut Core `1.2.1`. 
As soon as we get closer to a Micronaut Core `1.3.0` release I will start implementing the multipart support in the Micronaut GraphQL module as all building blocks are available now.

**Comment by marceloverdijk** [comment]
Putting this on hold to consider whether to add non-[official GraphQL spec](https://github.com/graphql/graphql-spec) related functionality or not.

**Comment by javacques** [comment]
Hi there!
For those looking for a solution, a very basic implementation of the spec can be found here - https://gist.github.com/javacques/725c9a574c3cf0dc7b66e2061c5f4eca.
You can add an Upload scalar to the `RuntimeWiring`:
```java
RuntimeWiring runtimeWiring = RuntimeWiring.newRuntimeWiring()
        .scalar(CompletedFileUploadScalar.UPLOAD_SCALAR)
        ...
        .build();
```
and then get `CompletedFileUpload` in your `DataFetcher` like this:
```java
CompletedFileUpload file = environment.getArgument("file");
```

**Comment by magnayn** [comment]
... driving by, 1.5 years after the above...

Would it not work to merge the above gist into the codebase? 

It seems a shame that it doesn't check for Multipart content..

**Comment by graemerocher** [comment]
I started working on this issue

**Comment by graemerocher** [comment]
I started working on this issue

**Comment by graemerocher** [comment]
I started working on this issue
